### PR TITLE
Extend enos-run.yml to cover E2E testing

### PR DESCRIPTION
**Pull request by manthonygfp**
This PR is related to adding a GitHub Action workflow that executes the Enos End to End scenario.

We have chosen to add on to the existing Enos Run workflow that also runs integration tests. In the process, we have inverted some logic that previously checked the affirmative case of "a" and now checks the negative case of "b". This was necessary to add more than 2 types of filters to the workflow.

Additionally, another deviation from previous implementations is that we will only run the Destroy step if either the Run or its retry step have failed. This is because `enos scenario run` will execute a destroy action if it completes its apply action successfully.

Other than that, the only other change is that we have added a new filter to the existing scenario in order to execute this new scenario.

### CODE REVIEW

File: .github/workflows/enos-run.yml
Comment (jaymalasinha) [inline .github/workflows/enos-run.yml:47]: if you need this step for both `cli_ui` and `e2e` scenarios but not for `smoke` you might be able to update this condition to `if: matrix.test != 'smoke'`

File: .github/workflows/enos-run.yml
Comment (jaymalasinha) [inline .github/workflows/enos-run.yml:161]: It seems even though the destroy step failed before it did not error out in GHA with Enos version 0.0.10 and before, but now failure in destroy step does return error with Enos 0.0.11 .  May be we can add `continue-on-error: true` as a temporary workaround to this step with a comment, as the failure of destroy step does not determine the failure of Enos run

**Review by jaymalasinha — COMMENTED** [review]
Just left couple comments, otherwise looks great 👍

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:161]: (Not replying to Jaymala here. We talked about this in Slack. This comment is for posterity.)

In the case of continue-on-error, there will still be an icon indicating this step failed if the runs were successful. We know that's okay, but that isn't immediately clear to a random user that may see this.

continue-on-error is present for the whole job which already has this suggested functionality present, but my desire is to not have the red X indicating failure on this step if the step didn't need to run in the first place.

This solution keeps the step from running when it doesn't need to run and shows that it was appropriately skipped. This allows the user to just glance at it and not need to investigate further or have tribal knowledge about the workflow.

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [inline .github/workflows/enos-run.yml:144]:
```suggestion
          enos scenario run --timeout 60m0s --chdir ./enos ${{ matrix.filter }} builder:crt
```

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [inline .github/workflows/enos-run.yml:159]:
```suggestion
          enos scenario run --timeout 60m0s --chdir ./enos ${{ matrix.filter }} builder:crt
```

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [inline .github/workflows/enos-run.yml:173]:
```suggestion
          enos scenario destroy --timeout 60m0s --chdir ./enos ${{ matrix.filter }} builder:crt
```

File: .github/workflows/enos-run.yml
Comment (moduli) [reply .github/workflows/enos-run.yml:161]: I'm a little confused.

The previous steps are executing `enos scenario run` which, based on my understanding, includes a `destroy`. Previously, this step executes another destroy. Now, this change is suggesting to only run this destroy if `run` or `run_retry` failed (whether it failed when running or destroying? Is that correct?

I'm not clear on what `continue-on-error` would do here.

Comment (manthonygfp) [reply .github/workflows/enos-run.yml:161]: Addendum to the last bit, for posterity: Jaymala suggested that this could and probably should be handled in Enos itself. Terraform _also_ is okay with saying "There is nothing to destroy. My job is done." I'm inclined to agree that a destroy that does not need to destroy anything should exit 0. Hopefully this is temporary, and we will add functionality in Enos to allow this to execute without condition.

File: .github/workflows/enos-run.yml
Comment (moduli) [inline .github/workflows/enos-run.yml:104]: In one of the new scenarios I'm adding, it will also need `jq`. I presume we'll need to add another block here for that?

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:43]: Oh great catch- this was a debug statement that was there to ensure that we were in fact using the same version of Go as the .go-version file (which it was, of course). I'll remove it.

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:161]: @moduli That's correct. The destroy part of the lifecycle of `enos scenario run` only executes if the apply part of it was successful. This discrete call to `enos scenario destroy` ensures that the infrastructure is always destroyed even in the case of the run or run-retry step's failure.

With the previous logic, if the destroy step ran after the run was successful, you would have an icon on the workflow run page that indicates the failure of a step. This change makes it to where the step is explicitly skipped in the event that the run (or retry of course) step was successful.

But, because of the discussion in this PR, Ryan is making a change to Enos that will make the `enos scenario destroy` action not return a non-zero exit code if it simply has nothing to destroy. As a result, we will likely move the conditional from this step before it is merged.

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:104]: You make a good point implicitly that this could grow to be unwieldy. I think one approach would be for us to make the inversion of this logic instead, but maybe we could more simply _always_ install `jq` as it won't harm scenarios that _don't_ use it.

**Comment by manthonygfp** [comment]
One of my recent commits has an inaccurate commit message, but these will be squashed so I'm not going to go through the effort of amending it.

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:104]:
```suggestion
        run: |
          chmod +x /tmp/bats-cli-ui-deps/jq-bin
          sudo cp /tmp/bats-cli-ui-deps/jq-bin /usr/local/bin/jq
```

What do we think about removing it?

File: enos/enos-variables.hcl
Comment (manthonygfp) [inline enos/enos-variables.hcl:82]: We could maybe do away with this as it was part of trying to figure out a previous issue. It's not harming anything to be here.
```suggestion
```

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [inline .github/workflows/enos-run.yml:43]:
```suggestion
```

File: .github/workflows/enos-run.yml
Comment (moduli) [reply .github/workflows/enos-run.yml:161]: Ah. Got it.

File: enos/enos-variables.hcl
Comment (moduli) [reply enos/enos-variables.hcl:82]: I would lean towards removing it if it's not used for anything.

File: .github/workflows/enos-run.yml
Comment (moduli) [reply .github/workflows/enos-run.yml:104]: Seems ok? Reminder: we'd have to remove it in the download step above, too.

Alternatively, would it make sense for me to handle this in the scenario? Would that even be possible?

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:47]: Ryan has updated Enos- I will remove the condition altogether 👏

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:101]: @jaymalasinha do you think that ensuring `jq` is present should be handled in the GHA workflow or Enos scenario?

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:158]: Removing the condition as Ryan has a change to Enos that will let destroy run and exit cleanly if it has nothing to destroy.

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [inline .github/workflows/enos-run.yml:158]:
```suggestion
```

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [reply .github/workflows/enos-run.yml:158]:
```suggestion
```

File: .github/workflows/enos-run.yml
Comment (jaymalasinha) [reply .github/workflows/enos-run.yml:101]: `jq` install can be done within the Enos scenario modules

Comment (manthonygfp) [reply .github/workflows/enos-run.yml:101]: Hm but do you think we should remove it here and edit all of the scenarios that need it to include it...

or should we delegate that task to the workflow?

I guess my question is that we don't really want to be doing this in different ways that get stuffed into the same workflow (some installing jq with Enos, some doing it with the GHA workflow).

Comment (manthonygfp) [reply .github/workflows/enos-run.yml:101]: I talked with Michael about this out of the comments- we're going to install it as part of the workflow as we already assume other local deps (like Go) are installed when this is running on someone's MacBook.

File: .github/workflows/enos-run.yml
Comment (manthonygfp) [inline .github/workflows/enos-run.yml:92]:
```suggestion
```

Comment (manthonygfp) [inline .github/workflows/enos-run.yml:92]:
```suggestion
        if: steps.dep-cache.outputs.cache-hit != 'true'
```

**Comment by manthonygfp** [comment]
I think we're good to go, waiting on the changes to Enos we had talked about to be merged before merging the PR.

File: .github/workflows/enos-run.yml
Comment (moduli) [inline .github/workflows/enos-run.yml:92]: nit: For these updated steps that aren't checking for `cli_ui` anymore, we may want to update the `name` so that it's not referring to `Bats CLI UI tests`

**Comment by moduli** [comment]
Also, just noticed that the PR status check for this is failing? What's going on there?

**Review by jaymalasinha — APPROVED** [review]
LGTM

**Review by jaymalasinha — APPROVED** [review]

**Review by moduli — APPROVED** [review]

**Review by joshbrand — APPROVED** [review]
🎉 🎉 🎉

**Pull request merged**
